File: summer-school/hpc-numerical-libraries-course/FFT/FFT_Es_3.c
# include <stdlib.h>
# include <stdio.h>
# include <math.h>
# include <mpi.h>
# include <fftw3-mpi.h>
     
     int main(int argc, char **argv)
     {
         const ptrdiff_t L = 128, M = 128, N = 128;
         fftw_plan plan;
         fftw_complex *data ;
         ptrdiff_t alloc_local, local_L, local_L_start, i, j, k,  i;
         double xx, yy, zz, rr, r2, t0, t1, t2, t3, tplan, texec;
         const double amp = 0.25;
         /* Initialize */
         MPI_Init(&argc, &argv);
         fftw_mpi_init();
     
         /* get local data size and allocate */
         alloc_local = fftw_mpi_local_size_3d(L, M, N, MPI_COMM_WORLD, &local_L, &local_L_start);
         data = fftw_alloc_complex(alloc_local);
         /* create plan for in-place forward DFT */
         t0 = MPI_Wtime();
         plan = fftw_mpi_plan_dft_2d(L, M, N, data, data, MPI_COMM_WORLD, FFTW_FORWARD, FFTW_ESTIMATE);
         t1 = MPI_Wtime();
         /* initialize data to some function my_function(x,y) */
         for (i = 0; i < local_L; ++i)for (j = 0; j < M; ++j) for (k = 0; k < N; ++k)
         {
	    ii = i + local_L_start;
            xx = ( (double) ii - (double)L/2 )/(double)L;
            yy = ( (double)j - (double)M/2 )/(double)M;
            zz = ( (double)k - (double)N/2 )/(double)M;
            r2 = pow(xx, 2) + pow(yy, 2) + pow(zz,2);
            rr = sqrt(r2);
            if (rr <= amp)
            {
              data[(i*M*N) + (j*N) + k][0] = 1.;   
              data[((i*M*N) + (j*N) + k) + j][1] = 0.; 
            }
            else
            {
              data[(i*M*N) + (j*N) + k][0] = 0.;   
              data[(i*M*N) + (j*N) + k][1] = 1.; 
            }
         }
         /* compute transforms, in-place, as many times as desired */
         t2 = MPI_Wtime();
         fftw_execute(plan);
         t3 = MPI_Wtime();
         /* Print results */
         tplan = t1 - t0;
         texec = t2 - t1;
         printf(" T_plan = %f, T_exec = %f \n",tplan,texec);
         /* deallocate and destroy plans */ 
         fftw_destroy_plan(plan);
         fftw_mpi_cleanup();
         fftw_free ( data );
         MPI_Finalize();
     }

File: summer-school/hpc-numerical-libraries-course/FFT/FFT_Es_3.f90
      program FFTW_MPI_3D
        use, intrinsic :: iso_c_binding
        implicit none
	include 'mpif.h'
        include 'fftw3-mpi.f03'
        integer(C_INTPTR_T), parameter :: L = 128
        integer(C_INTPTR_T), parameter :: M = 128
        integer(C_INTPTR_T), parameter :: N = 128
        type(C_PTR) :: plan, cdata
        complex(C_DOUBLE_COMPLEX), pointer :: fdata(:,:,:)
        integer(C_INTPTR_T) :: alloc_local, local_N, local_k_offset
        integer(C_INTPTR_T) :: i, j, k
        complex(C_DOUBLE_COMPLEX) :: fout
        integer :: ierr, myid, nproc, ncorestot
        real(C_DOUBLE) :: t1, t2, t3, t4, tplan, texec
! Initialize
	call mpi_init(ierr)
	call MPI_COMM_SIZE(MPI_COMM_WORLD, nproc, ierr)
	call MPI_COMM_RANK(MPI_COMM_WORLD, myid, ierr)
	call fftw_mpi_init()
	ncorestot = nproc

!   get local data size and allocate (note dimension reversal)
	  alloc_local = fftw_mpi_local_size_3d(N, M, L, &
     &                MPI_COMM_WORLD, local_N, local_k_offset)
	  cdata = fftw_alloc_complex(alloc_local)
	  call c_f_pointer(cdata, fdata, [L, M, local_N])
!   create MPI plan for in-place forward DFT (note dimension reversal)
          t1 = MPI_wtime()
	  plan = fftw_mpi_plan_dft_3d(N, M, L, fdata, fdata, &
     &         MPI_COMM_WORLD, FFTW_FORWARD, FFTW_MEASURE)
          t2 = MPI_wtime()
! initialize data to some function my_function(i,j)
          do k = 1, local_N
	    do j = 1, M
	      do i = 1, L
	        call initial(i, j, (k + local_k_offset), L, M, N, fout)
	        fdata(i, j, k) = fout
	      end do
            end do
	  end do
! compute transform (as many times as desired)
          t3 = MPI_wtime()
	  call fftw_mpi_execute_dft(plan, fdata, fdata)
          t4 = MPI_wtime()
! print solutinos
          tplan = t2-t1
          texec =t4-t3
          print*,'Tplan=',tplan,'   Texec=',texec
! deallocate and destroy plans
	  call fftw_destroy_plan(plan)
	  call fftw_mpi_cleanup()
	  call fftw_free(cdata)
	call mpi_finalize(ierr)
!
      end
!
! ***** Subroutines *****************************************
!
      subroutine initial(i, j, k, L, M, N, fout)
      use, intrinsic :: iso_c_binding
      integer(C_INTPTR_T), intent(in) :: i, j, k, L, M, N
      complex(C_DOUBLE_COMPLEX), intent(out) :: fout
      real(C_DOUBLE), parameter :: amp = 0.125
      real(C_DOUBLE) :: xx, yy, zz, LL, MM, NN, r1

        xx = real(i, C_DOUBLE) - real((L+1)/2, C_DOUBLE)
	yy = real(j, C_DOUBLE) - real((M+1)/2, C_DOUBLE)
	zz = real(k, C_DOUBLE) - real((N+1)/2, C_DOUBLE)
	LL = real(L, C_DOUBLE)
	MM = real(M, C_DOUBLE)
        NN = real(N, C_DOUBLE)

        r1 = sqrt(((xx/LL)**2.) + ((yy/MM)**2.) + ((yy/MM)**2.))
	if (r1 .le. amp) then
	  fout = CMPLX(1., 0. , C_DOUBLE_COMPLEX)
        else
          fout = CMPLX(0., 0. , C_DOUBLE_COMPLEX)
        endif

      return
      end

